File: app/src/main/java/com/johnpetitto/orachat/data/chat/ChatService.java
package com.johnpetitto.orachat.data.chat;

import com.johnpetitto.orachat.data.ApiResponse;

import java.util.List;

import io.reactivex.Single;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.PATCH;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface ChatService {
    @GET("chats")
    Single<ApiResponse<List<Chat>>> list(
            @Query("name") String name,
            @Query("page") int page,
            @Query("limit") int limit
    );

    @POST("chats")
    Single<ApiResponse<Chat>> create(@Body CreateChat createChat);

    @PATCH("chats/{id}")
    Single<ApiResponse<Chat>> update(@Path("id") long id, @Body String name);

    @GET("chats/{id}/chat_messages")
    Single<ApiResponse<List<ChatMessage>>> listMessages(
            @Path("id") long id,
            @Query("page") int page,
            @Query("limit") int limit
    );

    @POST("chats/{id}/chat_messages")
    Single<ApiResponse<ChatMessage>> createMessage(@Path("id") long id, @Body String message);
}

File: app/src/main/java/com/johnpetitto/orachat/data/chat/ChatService.kt
package com.johnpetitto.orachat.data.chat

import com.johnpetitto.orachat.data.ApiResponse
import io.reactivex.Single
import retrofit2.http.*

interface ChatService {
    @GET("chats")
    fun list(@Query("name") name: String?, @Query("page") page: Int, @Query("limit") limit: Int): Single<ApiResponse<List<Chat>>>

    @POST("chats")
    fun create(@Body chat: NewChat): Single<ApiResponse<Chat>>

    @PATCH("chats/{id}")
    fun update(@Path("id") id: Long, @Body name: String): Single<ApiResponse<Chat>>

    @GET("chats/{id}/chat_messages")
    fun listMessages(@Path("id") id: Long, @Query("page") page: Int, @Query("limit") limit: Int): Single<ApiResponse<List<ChatMessage>>>

    @POST("chats/{id}/chat_messages")
    fun createMessage(@Path("id") id: Long, @Body message: String): Single<ApiResponse<ChatMessage>>
}
